### PR TITLE
Refactor cmd execute

### PR DESCRIPTION
**Pull request by drewburlingame**
@bilal-fazlani  while reviewing the CommandLineApplication code, I found it difficult to grok all that was happening.  I made these changes as a way to clarify control flow and reduce duplicated logic.  I replaced `!processed` checks with `continue` statements to make it clear when items wouldn't be used again.  Cyclomatic complexity dropped from 145% to 115%.  It could be better still.  How would you feel if I submitted another PR extracting that logic into another class(es)?

### CODE REVIEW

File: CommandDotNet/MicrosoftCommandLineUtils/CommandLineApplication.cs [hunk 1]
```diff
@@ -154,101 +148,91 @@ public int Execute(params string[] args)
...
+                        optionParts = arg.Substring(2).Split(new[] { ':', '=' }, 2);
+                        var optionName = optionParts[0];
+                        option = command.FindOption(o => o.LongName, optionName);
 
                         if (option == null)
                         {
-                            if (string.IsNullOrEmpty(longOptionName) && !_appSettings.ThrowOnUnexpectedArgument  && _appSettings.AllowArgumentSeparator)
+                            var isArgumentSeparator = string.IsNullOrEmpty(optionName);
+                            
+                            // ??? if AllowArgumentSeparator is true, why would this be considered an unexpected argument?
```
Comment (drewburlingame) [inline CommandDotNet/MicrosoftCommandLineUtils/CommandLineApplication.cs]: @bilal-fazlani if AllowArgumentSeparator is true, why would this be considered an unexpected argument?

File: CommandDotNet/MicrosoftCommandLineUtils/CommandLineApplication.cs [hunk 2]
```diff
@@ -395,15 +378,30 @@ public string GetFullNameAndVersion()
...
-            }
-
+            var rootCmd = GetParentCommands().Last();
             Out.WriteLine(rootCmd.GetFullNameAndVersion());
             Out.WriteLine();
         }
+        
+        private IEnumerable<CommandLineApplication> GetSelfAndParentCommands()
+        {
+            for (CommandLineApplication c = this; c != null; c = c.Parent)
```
Comment (drewburlingame) [inline CommandDotNet/MicrosoftCommandLineUtils/CommandLineApplication.cs]: I liked this approach for looping through a hierarchy.  I think I'll use it from now on.  I'm not sure it's as immediately intuitive as a while statement but it's terser and the intent is obvious from the method it's in.

Comment (bilal-fazlani) [reply CommandDotNet/MicrosoftCommandLineUtils/CommandLineApplication.cs]: Nice. I find this more easy to understand

File: CommandDotNet/MicrosoftCommandLineUtils/CommandLineApplication.cs (hunk 1 above)
Comment (bilal-fazlani) [reply CommandDotNet/MicrosoftCommandLineUtils/CommandLineApplication.cs]: hmmm everything in the directory MicrosoftCommandLineUtils has been a black box for me. I copied this from a microsoft nuget package.

I will have to analyse this code to understand more.

Comment (drewburlingame) [reply CommandDotNet/MicrosoftCommandLineUtils/CommandLineApplication.cs]: This makes more sense to me, what do you think?

```c#
if (isArgumentSeparator && _appSettings.AllowArgumentSeparator)
{
    // skip over the '--' argument separator
    // all remaining arguments to be added to command.RemainingArguments 
    index++;
}

HandleUnexpectedArg(command, args, index, argTypeName: "option");
break;
```

Comment (bilal-fazlani) [reply CommandDotNet/MicrosoftCommandLineUtils/CommandLineApplication.cs]: Yes makes sense.

**Pull request merged**